File: integration_test/plugin_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:archive/archive.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:integration_test/integration_test.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:serious_python/serious_python.dart';

class TestTrack {
  final String name;
  final List<String> artists;
  final int durationMs;

  const TestTrack({
    required this.name,
    required this.artists,
    required this.durationMs,
  });

  Map<String, dynamic> toJson() => {
    'name': name,
    'artists': artists.map((a) => {'name': a}).toList(),
    'durationMs': durationMs,
  };
}

final List<TestTrack> realTestTracks = [
  const TestTrack(
    name: 'Come Together',
    artists: ['The Beatles'],
    durationMs: 259000, // 4:19
  ),
  const TestTrack(
    name: 'Bohemian Rhapsody',
    artists: ['Queen'],
    durationMs: 354000, // 5:54
  ),
];

Future<int> getUnusedPort() async {
  final socket = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
  final int port = socket.port;
  await socket.close();
  return port;
}

Future<void> waitForServerReady(
  int port, {
  Duration timeout = const Duration(seconds: 20),
}) async {
  final stopwatch = Stopwatch()..start();
  final client = http.Client();

  while (stopwatch.elapsed < timeout) {
    try {
      final response = await client.get(
        Uri.parse('http://127.0.0.1:$port/ping'),
      );
      if (response.statusCode == 200) {
        client.close();
        return;
      }
    } catch (_) {
      // Waiting for Python daemon to start
    }
    await Future.delayed(const Duration(milliseconds: 250));
  }

  client.close();
  throw TimeoutException(
    'Flask RPC server failed to start within ${timeout.inSeconds} seconds.',
  );
}

Future<void> verifyAudioStreamSource(Map<String, dynamic> stream) async {
  final url = stream['url'] as String?;
  expect(url, isNotNull);
  expect(url, startsWith('https://'));
  expect(url, contains('googlevideo.com'));

  final bitrate = stream['bitrate'] as num?;
  expect(bitrate, isNotNull);
  expect(bitrate, greaterThan(0));

  expect(stream['codec'], isNotNull);

  final expiresAt = stream['expiresAt'] as num?;
  expect(expiresAt, isNotNull);
  expect(expiresAt, greaterThan(DateTime.now().millisecondsSinceEpoch));

  final headers =
      (stream['headers'] as Map?)?.cast<String, String>() ?? <String, String>{};
  final requestHeaders = {...headers, 'Range': 'bytes=0-1024'};

  final client = http.Client();
  try {
    final response = await client.get(Uri.parse(url!), headers: requestHeaders);

    expect(
      [200, 206],
      contains(response.statusCode),
      reason:
          'Expected CDN response 200 or 206, but received HTTP ${response.statusCode}: ${response.reasonPhrase}',
    );

    final contentType = response.headers['content-type'] ?? '';
    expect(
      contentType,
      matches(RegExp(r'audio\/(mp4|webm|mpeg|opus)|video\/mp4')),
    );

    expect(response.bodyBytes.length, greaterThan(0));
    expect(response.bodyBytes.length, lessThanOrEqualTo(1025));
  } finally {
    client.close();
  }
}

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
  HttpOverrides.global = null;

  group('YouTube Audio Source Plugin Integration Tests', () {
    const sourceZipPath = 'plugin.zip';
    late Directory pluginInstallDir;
    late int serverPort;
    late http.Client httpClient;

    setUpAll(() async {
      final sourceZipFile = File(sourceZipPath);
      expect(
        sourceZipFile.existsSync(),
        isTrue,
        reason: 'plugin.zip must exist. Run "python tool/build.py" first!',
      );

      final appSupportDir = await getApplicationSupportDirectory();
      pluginInstallDir = Directory(
        p.join(
          appSupportDir.path,
          'plugins',
          'org.musicare.audiosource.youtube',
        ),
      );

      if (pluginInstallDir.existsSync()) {
        pluginInstallDir.deleteSync(recursive: true);
      }
      pluginInstallDir.createSync(recursive: true);

      final zipBytes = sourceZipFile.readAsBytesSync();
      final archive = ZipDecoder().decodeBytes(zipBytes);

      for (final file in archive) {
        final filename = p.join(pluginInstallDir.path, file.name);
        if (file.isFile) {
          final outFile = File(filename);
          outFile.createSync(recursive: true);
          outFile.writeAsBytesSync(file.content as List<int>);
        } else {
          Directory(filename).createSync(recursive: true);
        }
      }

      httpClient = http.Client();
      serverPort = await getUnusedPort();

      final entryPointPath = p.join(pluginInstallDir.path, 'src', 'main.py');
      Directory.current = pluginInstallDir;

      SeriousPython.runProgram(
        entryPointPath,
        modulePaths: [pluginInstallDir.path],
        environmentVariables: {'PORT': serverPort.toString()},
        sync: false,
      );

      await waitForServerReady(serverPort);
    });

    tearDownAll(() {
      httpClient.close();
      SeriousPython.terminate();

      if (pluginInstallDir.existsSync()) {
        try {
          pluginInstallDir.deleteSync(recursive: true);
        } catch (_) {}
      }
    });

    testWidgets('Health check responds with valid plugin metadata', (
      tester,
    ) async {
      final response = await httpClient.get(
        Uri.parse('http://127.0.0.1:$serverPort/ping'),
      );
      expect(response.statusCode, equals(200));

      final Map<String, dynamic> data = jsonDecode(response.body);
      expect(data['status'], equals('ready'));
      expect(data['id'], equals('org.musicare.audiosource.youtube'));
      expect(data['name'], equals('MusicAre YouTube Audio Source'));
    });

    for (final track in realTestTracks) {
      testWidgets(
        'Resolves and validates playable audio stream for "${track.artists.first} - ${track.name}"',
        (tester) async {
          final uri = Uri.parse('http://127.0.0.1:$serverPort/get_stream');

          final response = await httpClient.post(
            uri,
            headers: {'Content-Type': 'application/json'},
            body: jsonEncode({'track': track.toJson(), 'quality': 'high'}),
          );

          expect(
            response.statusCode,
            equals(200),
            reason: 'Server returned error: ${response.body}',
          );

          final dynamic data = jsonDecode(response.body);
          expect(data, isA<List>());

          final List<dynamic> sources = data as List;
          expect(sources, isNotEmpty);

          final Map<String, dynamic> primaryStream = Map<String, dynamic>.from(
            sources.first as Map,
          );
          await verifyAudioStreamSource(primaryStream);
        },
      );
    }
  });
}

File: test/plugin_test.py
from builtins import any
import unittest
import time
import urllib.request
from src.main import app
from src.sdk.types import Track, Artist
from src.matcher import TrackMatcher, CandidateTrack

class TestYouTubeAudioPlugin(unittest.TestCase):
    def setUp(self):
        """Initializes Flask test client before each test execution."""
        app.config['TESTING'] = True
        self.client = app.test_client()

    def test_health_check_returns_plugin_metadata(self):
        """Verifies that GET /ping returns 200 OK with correct plugin identity metadata."""
        response = self.client.get('/ping')
        self.assertEqual(response.status_code, 200)
        
        data = response.get_json()
        self.assertEqual(data['status'], 'ready')
        self.assertEqual(data['id'], 'org.musicare.audiosource.youtube')
        self.assertEqual(data['name'], 'MusicAre YouTube Audio Source')

    def test_matcher_ranks_candidates_by_adherence(self):
        """Verifies that TrackMatcher ranks candidates descending by duration and title score."""
        original = Track(
            name="Bohemian Rhapsody",
            artists=[Artist(name="Queen")],
            duration_ms=354000
        )

        candidates = [
            CandidateTrack(id="1", title="Bohemian Rhapsody (Live)", artist="Queen", duration_ms=420000),
            CandidateTrack(id="2", title="Bohemian Rhapsody - Metal Cover", artist="Random Band", duration_ms=354000),
            CandidateTrack(id="3", title="Bohemian Rhapsody (Official Audio)", artist="Queen", duration_ms=355000)
        ]

        ranked = TrackMatcher.rank_candidates(original, candidates)
        self.assertEqual(len(ranked), 3)
        self.assertEqual(ranked[0].id, "3")  # Official studio track must be ranked #1

    def test_resolve_stream_returns_playable_sources(self):
        """Verifies that POST /get_stream deciphers real stream URLs and CDN responds to HTTP Range requests."""
        test_tracks = [
            {
                "name": "Come Together",
                "artists": [{"name": "The Beatles"}],
                "durationMs": 259000
            },
            {
                "name": "Bohemian Rhapsody",
                "artists": [{"name": "Queen"}],
                "durationMs": 354000
            }
        ]

        for track_payload in test_tracks:
            with self.subTest(track=track_payload["name"]):
                response = self.client.post('/get_stream', json={
                    "track": track_payload,
                    "quality": "high"
                })

                self.assertEqual(response.status_code, 200)
                sources = response.get_json()

                self.assertIsInstance(sources, list)
                self.assertGreater(len(sources), 0)

                # Validate the primary resolved stream
                primary = sources[0]
                self.assertTrue(primary['url'].startswith('https://'))
                self.assertIn('googlevideo.com', primary['url'])
                self.assertGreater(primary['bitrate'], 0)
                self.assertIsNotNone(primary['codec'])
                self.assertGreater(primary['expiresAt'], int(time.time() * 1000))

                # Live HTTP Range handshake against Google CDN
                req = urllib.request.Request(
                    primary['url'],
                    headers={**(primary.get('headers') or {}), 'Range': 'bytes=0-1024'}
                )
                with urllib.request.urlopen(req, timeout=15) as cdn_response:
                    self.assertIn(cdn_response.status, [200, 206])
                    content_type = cdn_response.headers.get('Content-Type', '')
                    self.assertTrue(any(t in content_type for t in ['audio', 'video/mp4']))
                    self.assertGreater(len(cdn_response.read()), 0)

if __name__ == '__main__':
    unittest.main()